sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala

```diff
@@ -236,6 +236,58 @@ case class ShowFunctionsCommand(
   }
 }
 
+
+/**
+ * A command for users to refresh the persistent function.
+ * The syntax of using this command in SQL is:
+ * {{{
+ *    REFRESH FUNCTION functionName
+ * }}}
+ */
+case class RefreshFunctionCommand(
+    databaseName: Option[String],
+    functionName: String)
+  extends RunnableCommand {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val catalog = sparkSession.sessionState.catalog
+    if (FunctionRegistry.builtin.functionExists(FunctionIdentifier(functionName))) {
+      throw new AnalysisException(s"Cannot refresh native function $functionName")
+    } else if (catalog.isTemporaryFunction(FunctionIdentifier(functionName, databaseName))) {
+      throw new AnalysisException(s"Cannot refresh temp function $functionName")
+    } else {
+      // we only refresh the permanent function.
+      // there are 4 cases:
+      // 1. registry exists externalCatalog exists
+      // 2. registry exists externalCatalog not exists
+      // 3. registry not exists externalCatalog exists
+      // 4. registry not exists externalCatalog not exists
+      val identifier = FunctionIdentifier(
+        functionName, Some(databaseName.getOrElse(catalog.getCurrentDatabase)))
+      val isRegisteredFunction = catalog.isRegisteredFunction(identifier)
+      val isPersistentFunction = catalog.isPersistentFunction(identifier)
+      if (isRegisteredFunction && isPersistentFunction) {
+        // re-register function
+        catalog.unregisterFunction(identifier)
```

Comment:
We need `unregisterFunction` here even when `overwrite=true` in `registerFunction`?

Reply:
Since the `FunctionRegistry` is isolated across session, this will never happen in normal.

Reply:
Sorry, but probably I missed your point. If `catalog.unregisterFunction(identifier)` is removed, any test fails?

Reply:
We check `isRegisteredFunction` first so the function is in `FunctionRegistry` and `catalog.unregisterFunction(identifier)` always return true.

Reply:
Actually, I meant we cannot write it like this? It seems the session catalog already has the code for error handling such as `NoSuchFunctionException `.
```

  override def run(sparkSession: SparkSession): Seq[Row] = {
    val catalog = sparkSession.sessionState.catalog
    if (FunctionRegistry.builtin.functionExists(FunctionIdentifier(functionName))) {
      throw new AnalysisException(s"Cannot refresh builtin function $functionName")
    }
    if (catalog.isTemporaryFunction(FunctionIdentifier(functionName, databaseName))) {
      throw new AnalysisException(s"Cannot refresh temporary function $functionName")
    }

    val identifier = FunctionIdentifier(
      functionName, Some(databaseName.getOrElse(catalog.getCurrentDatabase)))
    val func = catalog.getFunctionMetadata(identifier)
    catalog.registerFunction(func, true)
  }
```

Reply:
We should check `isRegisteredFunction` and `isPersistentFunction` by self. 

Since the catalog may cache the function in `FunctionRegistry`, we should unregisterFunction then getFunctionMetadata and registerFunction. e.g. a function is not exists in hive but exists in cache.

Reply:
Ah, I see. I checked the code again and how about refactoring it like this (just a suggestion)?;
```
    val identifier = FunctionIdentifier(
      functionName, Some(databaseName.getOrElse(catalog.getCurrentDatabase)))
    if (catalog.isRegisteredFunction(identifier)) {
      // (...Write comments about why do we need to unregister a function...)
      // catalog.unregisterFunction(identifier)
    }
    if (catalog.isPersistentFunction(identifier)) {
      val func = catalog.getFunctionMetadata(identifier)
      catalog.registerFunction(func, true)
    } else {
      throw new NoSuchFunctionException(identifier.database.get, functionName)
    }
```
That's because many branches made me understood a bit hardly.